File: sdks/typescript/src/options.ts
// Copyright 2026 ConveyorQ
//
// SPDX-License-Identifier: Apache-2.0

import type { Encryptor } from "./encryption.js";
import type { Task } from "./task.js";

/**
 * TaskState is the lifecycle state of a task, as reported by
 * {@link Client.getTask}.
 */
export type TaskState =
  | "unspecified"
  | "scheduled"
  | "pending"
  | "active"
  | "retry"
  | "completed"
  | "archived"
  | "canceled"
  | "aggregating"
  | "blocked";

/**
 * DependencyFailure decides a dependent task's fate when a task it depends on
 * fails terminally (retries exhausted, skipped, or canceled) instead of
 * succeeding:
 * - `"block"` keeps the dependent blocked indefinitely (the default);
 * - `"cascade-cancel"` cancels the dependent and, in turn, its own dependents;
 * - `"continue"` treats the failed dependency as satisfied.
 */
export type DependencyFailure = "block" | "cascade-cancel" | "continue";

/**
 * Dependency is one task a task waits for. The dependent stays blocked until the
 * referenced task reaches a terminal success; `onFailure` decides what happens
 * if it fails terminally instead (defaults to `"block"`).
 */
export interface Dependency {
  /** The id of the task that must finish first. */
  taskId: string;
  /** Policy applied when the dependency fails terminally; defaults to "block". */
  onFailure?: DependencyFailure;
}

/** TaskInfo is the external view of a task returned by the producer API. */
export interface TaskInfo {
  id: string;
  queue: string;
  type: string;
  state: TaskState;
  priority: number;
  retried: number;
  maxRetry: number;
  lastError: string;
  enqueuedAt?: Date;
  processAt?: Date;
  completedAt?: Date;
  startedAt?: Date;
}

/**
 * EnqueueOptions configures one enqueue. Durations are in **milliseconds**;
 * absolute times are `Date`s. Every field is optional; unset fields take the
 * server default.
 */
export interface EnqueueOptions {
  /** A client-chosen task id, making enqueue retries idempotent. */
  taskId?: string;
  /** Target queue; defaults to "default". */
  queue?: string;
  /** Retries before the task is dead-lettered; 0 selects the server default. */
  maxRetry?: number;
  /** Dispatch priority within a queue, 1 (lowest) to 9 (highest). */
  priority?: number;
  /** Per-attempt timeout in milliseconds; the handler's signal aborts after it. */
  timeout?: number;
  /** Absolute time after which the task must not run. */
  deadline?: Date;
  /** Delay execution until this absolute time. */
  processAt?: Date;
  /** Delay execution by this many milliseconds. */
  processIn?: number;
  /** Keep the completed task visible for this many milliseconds before purge. */
  retention?: number;
  /** Reject duplicates of this task for this many milliseconds (uniqueness TTL). */
  unique?: number;
  /** Explicit uniqueness key; defaults to a hash of type + payload. */
  uniqueKey?: string;
  /** Make the task a member of the named aggregation group. */
  group?: string;
  /** Archive the task if it is not dispatched within this many milliseconds. */
  expiresIn?: number;
  /** Archive the task if it is not dispatched by this absolute time. */
  expiresAt?: Date;
  /**
   * Tasks this task waits for, building a workflow (a chain, or a fan-in
   * continuation). It stays blocked until each reaches a terminal success. A
   * plain string is a dependency with the default block-on-failure policy; a
   * {@link Dependency} sets an explicit policy. Dependencies must be acyclic.
   */
  dependsOn?: (string | Dependency)[];
}

/** EnqueueFn commits a task and returns its info; the unit enqueue middleware wraps. */
export type EnqueueFn = (task: Task, options: EnqueueOptions) => Promise<TaskInfo>;

/**
 * EnqueueMiddleware decorates the enqueue path, outermost first. It is the
 * client-side counterpart of {@link Mux.use}: inject metadata, enforce policy,
 * or record metrics before a task is committed.
 */
export type EnqueueMiddleware = (next: EnqueueFn) => EnqueueFn;

/** ClientOptions configures a {@link Client}. */
export interface ClientOptions {
  /** Bearer token presented to the server. */
  token?: string | undefined;
  /** Encrypt task payloads end to end before enqueue (see {@link newAESGCM}). */
  encryptor?: Encryptor | undefined;
  /** Middleware applied to every enqueue, outermost first. */
  enqueueMiddleware?: EnqueueMiddleware[] | undefined;
}

/** WorkerOptions configures a {@link Worker}. */
export interface WorkerOptions {
  /** Bearer token presented to the server. */
  token?: string | undefined;
  /** Queues this worker serves, mapping queue name to dispatch weight (> 0). */
  queues: Record<string, number>;
  /** Total simultaneous executions across all queues (> 0). */
  concurrency: number;
  /** Reported in Hello; defaults to the SDK's own version. */
  sdkVersion?: string | undefined;
  /** Minimum server version required, as semver (e.g. "v1.2.0"). */
  minServerVersion?: string | undefined;
  /** Decrypt encrypted payloads on dispatch (see {@link newAESGCM}). */
  encryptor?: Encryptor | undefined;
}

File: sdks/go/options.go
// Copyright 2026 ConveyorQ
//
// SPDX-License-Identifier: Apache-2.0

package conveyor

import (
	"time"

	"github.com/conveyorq/conveyor/encryption"
)

// Option configures a Client or a Worker.
type Option func(*options)

// options collects the settings shared by NewClient and NewWorker; the
// constructors validate what applies to them.
type options struct {
	// token is the bearer token presented to the server.
	token string
	// queues maps queue name to dispatch weight (workers only).
	queues map[string]int
	// concurrency is the worker's total execution slots (workers only).
	concurrency int
	// minServerVersion is the minimum server version the worker requires
	// (workers only); empty imposes no requirement.
	minServerVersion string
	// enqueueMiddleware decorates Client.Enqueue, outermost first (clients only).
	enqueueMiddleware []EnqueueMiddlewareFunc
	// encryptor encrypts task payloads end to end; nil leaves them in clear.
	encryptor encryption.Encryptor
}

// WithToken authenticates with the given bearer token.
func WithToken(token string) Option {
	return func(o *options) { o.token = token }
}

// WithQueues declares the queues a worker serves, mapping queue name to
// dispatch weight. Workers require at least one queue.
func WithQueues(queues map[string]int) Option {
	return func(o *options) { o.queues = queues }
}

// WithConcurrency sets the worker's total concurrent execution slots.
func WithConcurrency(n int) Option {
	return func(o *options) { o.concurrency = n }
}

// WithMinServerVersion requires the connected server to be at least the given
// semver version (e.g. "v1.2.0"). The server refuses the session if it is
// older. A non-semver value is ignored by the server's check; the empty
// default imposes no requirement.
func WithMinServerVersion(version string) Option {
	return func(o *options) { o.minServerVersion = version }
}

// WithEnqueueMiddleware appends middleware applied to every Client.Enqueue
// call, outermost first. It is the client-side counterpart of Mux.Use, letting
// callers inject metadata, enforce policy, or record metrics on the enqueue
// path. Passing a nil middleware panics.
func WithEnqueueMiddleware(middleware ...EnqueueMiddlewareFunc) Option {
	return func(o *options) {
		for _, wrap := range middleware {
			if wrap == nil {
				panic("conveyor: WithEnqueueMiddleware with nil middleware")
			}

			o.enqueueMiddleware = append(o.enqueueMiddleware, wrap)
		}
	}
}

// WithEncryption encrypts task payloads end to end with enc: a client seals
// each payload before it is enqueued, and a worker opens it on dispatch, so the
// server only ever stores and relays ciphertext — it holds no keys. Set the
// same Encryptor on every client and worker that share a queue. Use
// encryption.NewAESGCM for the built-in AES-256-GCM scheme, or supply your own
// Encryptor backed by a KMS, HSM, or custom codec.
//
// A nil enc is ignored, leaving encryption off, so callers can pass an
// optionally-configured Encryptor without a nil check.
//
// A worker decrypts only payloads that were sealed by an encrypting client, so
// encrypted and plaintext tasks may coexist on one queue; a worker that
// receives an encrypted task without an Encryptor fails the task rather than
// processing ciphertext.
func WithEncryption(enc encryption.Encryptor) Option {
	return func(o *options) { o.encryptor = enc }
}

// EnqueueOption configures one Enqueue call.
type EnqueueOption func(*enqueueOptions)

// enqueueOptions collects per-enqueue settings.
type enqueueOptions struct {
	// taskID is a client-assigned id for idempotent retries.
	taskID string
	// queue is the target queue; empty selects the server default.
	queue string
	// maxRetry is the retry budget; zero selects the server default.
	maxRetry int
	// priority orders dispatch within a queue; zero selects the default.
	priority int
	// timeout bounds a single execution attempt; zero leaves it unbounded.
	timeout time.Duration
	// deadline is the absolute time after which the task must not run.
	deadline time.Time
	// processAt delays execution to an absolute time.
	processAt time.Time
	// processIn delays execution by a duration.
	processIn time.Duration
	// expiresAt is the absolute time after which an undispatched task is
	// archived instead of run.
	expiresAt time.Time
	// expiresIn is the duration after enqueue after which an undispatched task
	// is archived instead of run.
	expiresIn time.Duration
	// retention keeps the completed task visible before purge.
	retention time.Duration
	// uniqueKey enforces uniqueness among incomplete tasks.
	uniqueKey string
	// uniqueTTL bounds how long the uniqueness claim is held.
	uniqueTTL time.Duration
	// group makes the task an aggregation-group member.
	group string
	// dependsOn lists the tasks this task waits for before it becomes eligible.
	dependsOn []Dependency
	// concurrencyKey caps how many tasks sharing it run at once on its queue.
	concurrencyKey string
	// retryStrategy overrides the server's default backoff growth; zero
	// (RetryDefault) keeps it.
	retryStrategy RetryStrategy
	// retryBase overrides the first-retry delay ceiling; zero keeps the default.
	retryBase time.Duration
	// retryMax overrides the overall retry delay cap; zero keeps the default.
	retryMax time.Duration
}

// RetryStrategy selects how a task's retry backoff delay grows with the attempt.
type RetryStrategy int

const (
	// RetryDefault uses the server's default backoff strategy.
	RetryDefault RetryStrategy = iota
	// RetryExponential doubles the delay ceiling each retry.
	RetryExponential
	// RetryLinear grows the delay ceiling linearly with the attempt.
	RetryLinear
	// RetryFixed holds the delay ceiling constant across retries.
	RetryFixed
)

// TaskID assigns a client-chosen task id, making Enqueue retries
// idempotent: re-enqueueing an existing id is a no-op.
func TaskID(id string) EnqueueOption {
	return func(o *enqueueOptions) { o.taskID = id }
}

// Queue routes the task to the named queue instead of "default".
func Queue(name string) EnqueueOption {
	return func(o *enqueueOptions) { o.queue = name }
}

// MaxRetry sets how many times a failing task is retried before archiving.
func MaxRetry(n int) EnqueueOption {
	return func(o *enqueueOptions) { o.maxRetry = n }
}

// Priority orders dispatch within a queue, 1 (lowest) to 9 (highest);
// unset tasks run at the default priority 4.
func Priority(p int) EnqueueOption {
	return func(o *enqueueOptions) { o.priority = p }
}

// Timeout bounds a single execution attempt: the handler's context is
// canceled after the duration. The effective deadline is the earliest of the
// timeout, any Deadline, and the lease expiry.
func Timeout(d time.Duration) EnqueueOption {
	return func(o *enqueueOptions) { o.timeout = d }
}

// Deadline sets an absolute time after which the task must not run; the
// handler's context is canceled at that time if execution is still in flight.
func Deadline(t time.Time) EnqueueOption {
	return func(o *enqueueOptions) { o.deadline = t }
}

// ProcessAt delays execution until the given time.
func ProcessAt(t time.Time) EnqueueOption {
	return func(o *enqueueOptions) { o.processAt = t }
}

// ProcessIn delays execution by the given duration.
func ProcessIn(d time.Duration) EnqueueOption {
	return func(o *enqueueOptions) { o.processIn = d }
}

// ExpiresAt drops the task if it has not been dispatched by the given time:
// a task still waiting then is archived instead of run. Use it for work that
// loses its value once a moment passes (a one-time code, a timely
// notification). It is distinct from Deadline, which cancels a task already
// running, and Retention, which purges a task already completed. Mutually
// exclusive with ExpiresIn.
func ExpiresAt(t time.Time) EnqueueOption {
	return func(o *enqueueOptions) { o.expiresAt = t }
}

// ExpiresIn drops the task if it has not been dispatched within the given
// duration of enqueue; it is the relative form of ExpiresAt, resolved against
// the server clock. Mutually exclusive with ExpiresAt.
func ExpiresIn(d time.Duration) EnqueueOption {
	return func(o *enqueueOptions) { o.expiresIn = d }
}

// Retention keeps the completed task row visible for inspection for the
// given duration before it is purged.
func Retention(d time.Duration) EnqueueOption {
	return func(o *enqueueOptions) { o.retention = d }
}

// Unique rejects the enqueue with ErrDuplicateTask while an incomplete
// task with the same uniqueness key exists, for at most ttl. The key is
// derived from the task type and payload; combine with UniqueKey to choose
// it explicitly.
func Unique(ttl time.Duration) EnqueueOption {
	return func(o *enqueueOptions) { o.uniqueTTL = ttl }
}

// UniqueKey sets the explicit uniqueness key compared by Unique, e.g.
// "user:42:welcome", instead of the derived type-and-payload key.
func UniqueKey(key string) EnqueueOption {
	return func(o *enqueueOptions) { o.uniqueKey = key }
}

// ConcurrencyKey caps how many tasks sharing the key run at once on the task's
// queue: the queue dispatches at most its configured concurrency limit of tasks
// with this key simultaneously, holding the rest pending until an active one
// finishes (e.g. "customer:42" to bound in-flight work per customer). It has no
// effect unless the queue has a concurrency limit set. Mutually exclusive with
// Group.
func ConcurrencyKey(key string) EnqueueOption {
	return func(o *enqueueOptions) { o.concurrencyKey = key }
}

// RetryPolicy overrides the server's default retry backoff for this task: the
// growth strategy plus the first-retry delay (base) and the overall cap (max).
// A zero base or max keeps the server default for that field, and RetryDefault
// keeps the server's strategy, so you can override only the part you need, e.g.
// RetryPolicy(RetryFixed, time.Minute, time.Minute) for a steady one-minute
// retry against a rate-limited downstream.
func RetryPolicy(strategy RetryStrategy, base, maxDelay time.Duration) EnqueueOption {
	return func(o *enqueueOptions) {
		o.retryStrategy = strategy
		o.retryBase = base
		o.retryMax = maxDelay
	}
}
